src/marqo/core/typed_vespa_index.py

```diff
@@ -0,0 +1,271 @@
+from typing import Dict, Any, List
+
+from marqo.core.models import MarqoQuery, MarqoIndex
+from marqo.core.models.marqo_index import FieldType, FieldFeature, DistanceMetric, IndexType
+from marqo.core.vespa_index import VespaIndex
+from marqo.exceptions import InvalidArgumentError
+from marqo.s2_inference import s2_inference
+from marqo.s2_inference.errors import UnknownModelError
+
+
+class TypedVespaIndex(VespaIndex):
+    _TYPE_MAP = {
+        FieldType.Text: 'string',
+        FieldType.Bool: 'bool',
+        FieldType.Int: 'int',
+        FieldType.Float: 'float',
+        FieldType.ArrayText: 'array<string>',
+        FieldType.ArrayInt: 'array<int>',
+        FieldType.ArrayFloat: 'array<float>',
+        FieldType.ImagePointer: 'string'
+    }
+
+    _DISTANCE_METRIC_MAP = {
+        DistanceMetric.PrenormalizedAnguar: 'prenormalized-angular'
+    }
+
+    _INDEX_FIELD_PREFIX = 'marqo__lexical_'
+    _FILTER_FIELD_PREFIX = 'marqo__filter_'
+    _CHUNKS_FIELD_PREFIX = 'marqo__chunks_'
+    _EMBEDDING_FIELD_PREFIX = 'marqo__embeddings_'
+
+    _SCORE_MODIFIERS_FIELD = 'marqo__score_modifiers'
+
+    _BM25_RANK_PROFILE = 'bm25'
+    _EMBEDDING_SIMILARITY_RANK_PROFILE = 'embedding_similarity'
+    _MODIFIERS_RANK_PROFILE = 'modifiers'
+
+    _RANK_INPUT_QUERY_EMBEDDING = 'query_embedding'
+    _RANK_INPUT_MULT_WEIGHTS = 'mult_weights'
+    _RANK_INPUT_ADD_WEIGHTS = 'add_weights'
+
+    @classmethod
+    def generate_schema(cls, marqo_index: MarqoIndex) -> str:
+        cls._validate_index_type(marqo_index)
+
+        schema = []
+
+        schema.append(f'schema {marqo_index.name} {{')
+        cls._generate_document(marqo_index, schema)
+        cls._generate_rank_profiles(marqo_index, schema)
+        cls._generate_default_fieldset(marqo_index, schema)
+        cls._generate_summaries(marqo_index, schema)
+        schema.append('}')
+
+        return '\n'.join(schema)
+
+    @classmethod
+    def to_vespa_document(cls, marqo_document: Dict[str, Any], marqo_index: MarqoIndex) -> Dict[str, Any]:
+        pass
+
+    @classmethod
+    def to_marqo_document(cls, vespa_document: Dict[str, Any], marqo_index: MarqoIndex) -> Dict[str, Any]:
+        pass
+
+    @classmethod
+    def to_vespa_query(cls, query: MarqoQuery) -> Dict[str, Any]:
+        pass
+
+    @classmethod
+    def _generate_document(cls, marqo_index: MarqoIndex, schema: List[str]) -> None:
```

Comment:
Is it possible to keep this method as pure as possible? So not mutating `schema`, and instead returning a new one? I am not a huge fan of these mutating-style methods

Reply:
Same as above